control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go

```diff
@@ -2862,6 +2928,128 @@ func (r *reconciler) reconcileAzureCloudNodeManager(ctx context.Context, image s
 	return errs
 }
 
+// reconcileGlobalPullSecret reconciles the original pull secret given by HCP and merges it with a new pull secret provided by the user.
+// The new pull secret is only stored in the DataPlane side so, it's not exposed in the API. It lives in the kube-system namespace of the DataPlane.
+// If that PS exists, the HCCO deploys a DaemonSet which mounts the whole Root FS of the node, and merges the new PS with the original one.
+// If the PS doesn't exist, the HCCO doesn't do anything.
+func (r *reconciler) reconcileGlobalPullSecret(ctx context.Context, hcp *hyperv1.HostedControlPlane, cpoImage string) error {
+	var (
+		userProvidedPullSecretBytes []byte
+		originalPullSecretBytes     []byte
+		globalPullSecretBytes       []byte
+		err                         error
+	)
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling global pull secret")
+
+	// Get the user provided pull secret
+	exists, additionalPullSecret, err := globalpullsecret.AdditionalPullSecretExists(ctx, r.kubeSystemSecretClient)
+	if err != nil {
+		return fmt.Errorf("failed to check if user provided pull secret exists: %w", err)
+	}
+
+	// Reconcile the RBAC for the Global Pull Secret
+	if err := globalpullsecret.ReconcileGlobalPullSecretRBAC(ctx, r.uncachedClient, r.CreateOrUpdate, hcp.Namespace); err != nil {
+		return fmt.Errorf("failed to reconcile global pull secret RBAC: %w", err)
+	}
+
+	if !exists {
+		// Early cleanup
+		if recoverBeforeShutdown {
```

Comment:
what's the point of recoverBeforeShutdown? its value will be true every reconciliation loop